x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RRFRankCoordinatorContext.java

```diff
@@ -0,0 +1,164 @@
+            }
+        }
+
+        // return early if we have no valid results
+        if (queues.isEmpty()) {
+            return SortedTopDocs.EMPTY;
+        }
+
+        // rank the global doc sets using RRF from the previously
+        // built priority queues
+        // the score is calculated on-the-fly where we update the
+        // score if we already saw it as part of a previous query's
+        // doc set, otherwise we make a new doc and calculate the
+        // initial score
+        Map<RankKey, RRFRankDoc> results = Maps.newMapWithExpectedSize(queryCount * windowSize);
+        final int fqc = queryCount;
+        for (int qi = 0; qi < queryCount; ++qi) {
+            PriorityQueue<RRFRankDoc> queue = queues.get(qi);
+            final int fqi = qi;
+            for (int rank = queue.size(); rank > 0; --rank) {
+                RRFRankDoc rrfRankDoc = queue.pop();
+                final int frank = rank;
+                results.compute(new RankKey(rrfRankDoc.doc, rrfRankDoc.shardIndex), (key, value) -> {
+                    if (value == null) {
+                        value = new RRFRankDoc(rrfRankDoc.doc, rrfRankDoc.shardIndex, fqc);
+                    }
+
+                    value.score += 1.0f / (rankConstant + frank);
+                    value.positions[fqi] = frank - 1;
+                    value.scores[fqi] = rrfRankDoc.scores[fqi];
+
+                    return value;
+                });
+            }
+        }
+
+        // sort the results based on rrf score, tiebreaker based on smaller shard then smaller doc id
+        RRFRankDoc[] sortedResults = results.values().toArray(RRFRankDoc[]::new);
+        Arrays.sort(sortedResults, (RRFRankDoc rrf1, RRFRankDoc rrf2) -> {
+            if (rrf1.score != rrf2.score) {
+                return rrf1.score < rrf2.score ? 1 : -1;
+            }
+            if (rrf1.shardIndex != rrf2.shardIndex) {
+                return rrf1.shardIndex < rrf2.shardIndex ? -1 : 1;
+            }
+            return rrf1.doc < rrf2.doc ? -1 : 1;
+        });
+        RRFRankDoc[] topResults = new RRFRankDoc[Math.min(size, sortedResults.length - from)];
+        for (int rank = 0; rank < topResults.length; ++rank) {
+            topResults[rank] = sortedResults[rank];
+            topResults[rank].rank = rank + 1 + from;
+        }
+        // update fetch hits for the fetch phase, so we gather any additional
+        // information required just like a standard query
+        assert topDocsStats.fetchHits == 0;
+        topDocsStats.fetchHits = topResults.length;
+
+        // return the top results where sort, collapse fields,
+        // and completion suggesters are not allowed
+        return new SortedTopDocs(topResults, false, null, null, null, 0);
+    }
+
+    public void decorateSearchHit(ScoreDoc scoreDoc, SearchHit searchHit) {
+        assert scoreDoc instanceof RRFRankDoc;
+        RRFRankDoc rankResult = (RRFRankDoc) scoreDoc;
+        searchHit.setRank(rankResult.rank);
```

Comment:
not too sure about positions (I asked above what it holds), but rank could be seen as something general enough for every ranker? Double checking that we need this method, maybe we could instead get (from RankDoc) and set the rank directly where this method is currently called?

Reply:
Moved rank to `RankDoc` and removed `decorateSearchHit`.